man/pssst.man
.\" Manpage for pssst.
.\" Contact christian@uhsat.de to correct errors or typos.
.TH man 1 "2017" "Pssst" "Pssst Manual"
.SH NAME
pssst \- Pssst CLI
.SH SYNOPSIS
pssst [option|command] [~|username:password@server] [receiver message...]
.SH DESCRIPTION
Pssst is a simple and secure way to exchange information.
.SH OPTIONS
.IP "-h, --help"
Show usage
.IP "-l, --license"
Show license
.IP "-v, --version"
Show version
.SH COMMANDS
.B create
.RS
Create user.
.RE
.B delete
.RS
Delete user.
.RE
.B pull
.RS
Pull messages.
.RE
.B push
.RS
Push message.
.RE
.SH FILES
.I ~/.pssst
.RS
Default user profile. See
.I README.md
for further details.
.RE
.I ~/.pssst.<user>
.RS
User specific data including own and cached keys.
.RE
.SH AUTHORS
Christian Uhsat <christian@uhsat.de>


CHANGELOG
2.6.4
* Changed Heroku app specification

2.6.3
* Fixed gitignore
* Limit max request size to 1 MB
* Changed CLI logo
* Changed READMEs
* Changed server debug output
* Changed default server port
* Changed fuzzy unit test size

2.6.2
* Fixed CLI user name bug

2.6.1
* Fixed server public key path
* Fixed documentation links

2.6.0
* Added Heroku config vars to `app.json`
* Fixed Heroku debug level config var
* Fixed server running warning
* Changed README locations
* Changed CLI logo
* Changed CLI password prompt
* Changed profile unit tests
* Security improvements
* Refactored server for performance

2.5.1
* Added graceful exit to server
* Added general server error handlerss
* Added stack trace to server errors
* Added Heroku debug flag `PSSST_DEBUG`
* Fixed documentation
* Refactored server error handling

2.5.0
* Changed header to `x-pssst-hash`
* Changed server error handling

2.4.5
* Fixed unit test
* Changed version link

2.4.4
* Added server support for Heroku keys
* Added client API key pinning again
* Fixed client documentation
* Changed to Python naming conventions
* Refactored HTTP server creation
* Refactored CLI key handling

2.4.3
* Added shield.io badges
* Changed all READMEs
* Removed API key pinning

2.4.2
* Added Heroku deploy link

2.4.1
* Fixed version number

2.4.0
* Fixed user name hashing
* Changed message structure
* Refactored tests

2.3.0
* Updated Redis dependency

2.2.0
* Fixed user name hashing
* Fixed API documentation
* Change message structure

2.1.0
* Added user name hashing
* Fixed API documentation
* Fixed CLI documentation
* Fixed password notice
* Fixed init exceptions
* Fixed typo

2.0.0
* Added default user profile
* Added command line shortcut
* Added help option to server
* Added Heroku deployment
* Changed Heroku add-on support
* Changed default quota to 1 MB
* Changed comments
* Changed API documentation
* Changed API address
* Changed API version
* Changed API bypass
* Changed READMEs
* Changed unit tests
* Changed manpage
* Changed gitignores
* Changed contact email address
* Removed startup verification
* Removed server sample config
* Removed server public folder
* Removed server time
* Removed metadata
* Removed shell mode from CLI
* Removed support for local HTTPS
* Removed support for old versions
* Removed Git attributes file
* Removed RFCs
* Removed GUI
* Hardened server
* Refactored tests
* Refactored client
* Refactored server heavily
* Refactored documentations
* Refactored Git ignore files

1.0.1
* Updated the servers key fingerprint

1.0.0
* Release (stable)

0.2.44
* Added Python requirement files
* Added support for IE 8 to GUI
* Added nvm to Travis CI script
* Added READMEs to sources
* Added GUI theme bootstrap-flat
* Added GUI responsive overrides
* Added GUI moment.js locale
* Fixed GUI HTML validation
* Fixes server usage text
* Fixed typo in comment
* Fixed Travis-CI script
* Moved server shell to scripts
* Changed CLI unit tests
* Changed GUI layout
* Changed README.md
* Changed ursa to node-rsa
* Updated Bower versions
* Updated NPM versions
* Updated Node.js to LTS version
* Updated Travis CI Node.js version
* Updated Travis CI format
* Renamed some server files
* Renamed some project folders
* Refactored project structure
* Refactored documentation
* Refactored server app
* Refactored GUI layout
* Refactored GUI logic

0.2.43
* Added all relevant RFC documents
* Moved documentation to doc folder
* Changed server sample config
* Changed limit to quota
* Changed README.md

0.2.42
* Added API RAML specification
* Added more Python versions to Travis CI
* Fixed padding specification to PKCS#7
* Fixed padding encoding bug
* Moved API specification to server folder
* Changed AES cipher to use CBC mode with PKCS#5 padding
* Changed test vectors script
* Changed README.md

0.2.41
* Added server maintenance shell
* Fixed some server code typos

0.2.40
* Changed RSA hash algorithm to SHA256
* Changed RSA key size to 2048 bit
* Changed public key fingerprint
* Changed test vectors script
* Changed README.md
* Removed reduced key size for unit testing

0.2.39
* Added the default user database limit
* Added a per user database entry limit
* Changed server user name testing logic
* Changed server box name testing logic
* Changed server config.json layout
* Changed server config.json.sample
* Removed name restriction from server
* Removed name restriction unit test
* Refactored server config handling

0.2.38
* Added scaling test script
* Added reply function to GUI
* Fixed bug in test.py script
* Fixed server command line parsing
* Fixed optimize bug in makeexe script
* Changed CLI unit test command line handling
* Changed CLI/GUI usage text
* Changed CLI error handling
* Changed CLI default encoding
* Changed example scripts error handling
* Updated Travis CI Python version
* Updated README.md
* Renamed test vectors script

0.2.37
* Refactored server
* Refactored user file structure
* Added support for multi API users
* Added Moment.js date parser to GUI
* Added link to official site to GUI
* Changed key names to UNIX standard
* Changed package.json versions
* Changed bower.json versions
* Changed copyright notice
* Changed unit tests
* Changed gitignore file
* Renamed makeexe script

0.2.36
* Added Windows PE build script
* Fixed key bug in Heroku script
* Fixed user name bug in GUI
* Changed CLI documentation
* Changed CLI parameters
* Changed GUI favicon
* Changed README.md
* Changed unit tests
* Renamed name token all
* Renamed class KeyStore

0.2.35
* Fixed bug with non ASCII characters
* Fixed example pull script box bug

src/cli/pssst.py
#!/usr/bin/env python
"""
Copyright (C) 2013-2015  Christian & Christian <hello@pssst.name>
Copyright (C) 2015-2017  Christian Uhsat <christian@uhsat.de>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.
"""
import binascii
import base64
import io
import json
import os
import re
import sys
import time

from getpass import getpass
from zipfile import ZipFile


try:
    from requests import request
    from requests.exceptions import ConnectionError, Timeout
except ImportError:
    sys.exit("Requires Requests (https://github.com/kennethreitz/requests)")


try:
    from Crypto import Random
    from Crypto.Cipher import AES, PKCS1_OAEP
    from Crypto.Hash import HMAC, SHA256
    from Crypto.Protocol.KDF import scrypt
    from Crypto.PublicKey import RSA
    from Crypto.Signature import PKCS1_v1_5
    from Crypto.Util.py3compat import bchr, bord, tobytes
except ImportError:
    sys.exit("Requires PyCrypto (https://github.com/dlitz/pycrypto)")


__all__, __version__ = ["Pssst", "CLI"], "2.12.1"


def _hexlify(data): # Utility shortcut
    return binascii.hexlify(data).decode("utf-8")


def _encode(data): # Utility shortcut
    return base64.b64encode(data).decode("utf-8")


def _decode(data): # Utility shortcut
    return base64.b64decode(data.encode("utf-8"))


class Pssst:
    """
    Pssst API low level communication class.

    Methods
    -------
    create()
        Creates an user.
    delete()
        Deletes an user.
    find(user)
        Returns the public key of an user.
    pull()
        Pulls all messages from the box.
    push(user, data)
        Pushes a message into the box.

    """
    class _User:
        """
        Internal parser class for canonical user names.

        Notes
        -----
        This class is not meant to be called externally.

        """
        def __init__(self, username, password=None, server=None):
            """
            Initializes the instance with the parsed user name.

            Parameters
            ----------
            param username : string
                User name (full or partial).
            param password : string, optional (default is None)
                User private key password.
            param server : string, optional (default is None)
                Server address.

            """
            username = username.strip()

            if not re.match("^(pssst\.)?\w{2,63}(:\S+)?(@\S+)?$", username):
                raise Exception("User name invalid")

            if username.startswith("pssst."):
                username = username[6:]

            if "@" in username and not server:
                username, server = username.split("@", 1)

            if ":" in username and not password:
                username, password = username.split(":", 1)

            self.name = username.lower()
            self.hash = _hexlify(scrypt(repr(self), b"[Pssst!]", 32, 16384, 8, 1, 1))
            self.profile = (self.name, password, server)

        def __repr__(self):
            """
            Returns the full user name in canonical notation.

            """
            return str("pssst.%s" % self.name)


    class _Key:
        """
        Internal key class providing cryptographic methods.

        Methods
        -------
        private(password)
            Returns the users private key (PEM format).
        public()
            Returns the users public key (PEM format).
        encrypt(data)
            Returns the encrypted data and nonce.
        decrypt(data, nonce)
            Returns the decrypted data.
        sign(data)
            Returns the data timestamp and signature.
        verify(data, timestamp, signature)
            Returns if data could be verified with timestamp and signature.

        Notes
        -----
        This class is not meant to be called externally.

        """
        RSA_SIZE, NONCE_SIZE, GRACE_TIME = 2048, 32 + AES.block_size, 5

        def __init__(self, key=None, password=None):
            try:
                if key:
                    self.key = RSA.importKey(key, password)
                else:
                    self.key = RSA.generate(Pssst._Key.RSA_SIZE)

            except (IndexError, TypeError, ValueError) as ex:
                raise Exception("Password wrong")

        def __epoch(self):
            return int(round(time.time()))

        def private(self, password):
            return self.key.exportKey("PEM", password, 8).decode("ascii")

        def public(self):
            return self.key.publickey().exportKey("PEM").decode("ascii")

        def encrypt(self, data):
            nonce = Random.get_random_bytes(Pssst._Key.NONCE_SIZE)
            size = AES.block_size - (len(data) % AES.block_size)
            data = tobytes(data) + (bchr(size) * size)

            data = AES.new(nonce[:32], AES.MODE_CBC, nonce[32:]).encrypt(data)
            nonce = PKCS1_OAEP.new(self.key).encrypt(nonce)

            return (data, nonce)

        def decrypt(self, data, nonce):
            nonce = PKCS1_OAEP.new(self.key).decrypt(nonce)
            data = AES.new(nonce[:32], AES.MODE_CBC, nonce[32:]).decrypt(data)

            return data[:-bord(data[-1])]

        def sign(self, data):
            current, data = self.__epoch(), data.encode("utf-8")

            hmac = HMAC.new(str(current).encode("ascii"), data, SHA256)
            hmac = SHA256.new(hmac.digest())

            signature = PKCS1_v1_5.new(self.key).sign(hmac)

            return (current, signature)

        def verify(self, data, timestamp, signature):
            current, data = self.__epoch(), data.encode("utf-8")

            hmac = HMAC.new(str(timestamp).encode("ascii"), data, SHA256)
            hmac = SHA256.new(hmac.digest())

            if abs(timestamp - current) <= Pssst._Key.GRACE_TIME:
                return PKCS1_v1_5.new(self.key).verify(hmac, signature)


    class _KeyStorage:
        """
        Internal storage class for public and private keys.

        Methods
        -------
        delete()
            Deletes the users key storage.
        server()
            Saves the public server key.
        list()
            Returns an alphabetical list of all key entries.
        load(entry)
            Returns a key.
        save(entry, key)
            Saves a key.

        Notes
        -----
        This class is not meant to be called externally.

        """
        def __init__(self, api, user, password):
            self.scheme = "%s"
            self.user = user
            self.file = os.path.join(os.path.expanduser("~"), repr(self))

            if os.path.exists(self.file):
                self.key = Pssst._Key(self.load("id_rsa"), password)
            else:
                self.key = Pssst._Key()
                self.save("id_rsa", self.key.private(password))

            self.scheme = re.sub("^(?i)https?://(.+)", "\g<1>/%s.pub", api)

            if "id_rsa" in self.list():
                self.api = Pssst._Key(self.load("id_rsa"))
            else:
                self.api = None

        def __repr__(self):
            return ".pssst." + self.user

        def __bool__(self):
            return self.__nonzero__()

        def __nonzero__(self):
            return os.path.exists(self.file)

        def delete(self):
            os.remove(self.file)

        def server(self, key):
            self.save("id_rsa", key)
            self.api = Pssst._Key(self.load("id_rsa"))

        def list(self):
            with ZipFile(self.file, "r") as file:
                keys = []

                # Filter out APIs
                for key in file.namelist():
                    if key.startswith(self.scheme.rsplit("/")[0]):
                        keys.append(re.sub("^.+/(.+)\.pub$", "\g<1>", key))

                return keys

        def load(self, entry):
            with ZipFile(self.file, "r") as file:
                return file.read(self.scheme % entry)

        def save(self, entry, key):
            with ZipFile(self.file, "a") as file:
                file.writestr(self.scheme % entry, key)


    def __init__(self, username, password, server=None):
        """
        Initializes the instance with an user object.

        Parameters
        ----------
        param username : string
            User name.
        param password : string
            User private key password.
        param server : string, optional (default is None)
            Server address.

        Raises
        ------
        Exception
            Because the username is required.
        Exception
            Because the password is required.

        Notes
        -----
        If the environment variable 'PSSST' exists, it will be used as the API
        address and port. If a server is given, it will override the API.

        """
        API = "http://localhost:62221"

        if not username:
            raise Exception("Username required")

        if not password:
            raise Exception("Password required")

        self.api = server or os.environ.get("PSSST", API)
        self.user = Pssst._User(username)
        self.keys = Pssst._KeyStorage(self.api, self.user.name, password)

        if not self.keys.api:
            self.keys.server(self.__request_url("key"))

    def __repr__(self):
        """
        Returns the client identifier.

        Returns
        -------
        string
            The client identifier.

        """
        return "Pssst CLI"

    def __request_api(self, method, path, data=None, auth=True):
        """
        Returns the result of an API request (signed and verified).

        Parameters
        ----------
        param method : string
            Request method.
        param path : string
            Request path.
        param data : JSON, optional (default is None)
            Request data.
        param auth : bool
            Request authentication.

        Returns
        -------
        string
            The response body.

        Raises
        ------
        Exception
            Because the user was deleted.
        Exception
            Because the signature is missing.
        Exception
            Because the signature is corrupt.
        Exception
            Because the verification has failed.

        Notes
        -----
        Please see the __init__ method.

        """
        if not self.keys:
            raise Exception("User was deleted")

        url = "%s/2/%s" % (self.api, path)
        body = str(json.dumps(data, separators=(",", ":"))) if data else ""
        headers = {
            "content-type": "application/json" if data else "text/plain",
            "user-agent": repr(self)
        }

        if auth:
            timestamp, signature = self.keys.key.sign(body)
            timestamp, signature = str(timestamp), _encode(signature)

            headers["x-pssst-hash"] = "%s; %s" % (timestamp, signature)

        response = request(method, url=url, data=body, headers=headers)

        mime = response.headers.get("content-type", "text/plain")
        head = response.headers.get("x-pssst-hash")
        body = response.text

        if not re.match("^[0-9]+; ?[A-Za-z0-9\+/]+=*$", head):
            raise Exception("Verification failed")

        timestamp, signature = head.split(";", 1)
        timestamp, signature = int(timestamp), _decode(signature)

        if not self.keys.api.verify(body, timestamp, signature):
            raise Exception("Verification failed")

        if response.status_code not in [200, 204]:
            raise Exception(body)

        if mime.startswith("application/json"):
            body = response.json()

        return body

    def __request_url(self, path):
        """
        Returns the result of an URL request (without any checks).

        Parameters
        ----------
        param path : string
            Requested path.

        Returns
        -------
        string
            The response body.

        Raises
        ------
        ConnectionError
            Because the file was not found.

        Notes
        -----
        Please see the __init__ method.

        """
        url = "%s/%s" % (self.api, path)
        headers = {
            "user-agent": repr(self)
        }

        response = request("GET", url=url, headers=headers)

        if response.status_code not in [200, 204]:
            raise ConnectionError("Not Found")

        return response.text

    def create(self):
        """
        Creates an user.

        """
        body = {"key": self.keys.key.public()}

        self.__request_api("POST", self.user.hash, body)

    def delete(self):
        """
        Deletes an user.

        Notes
        -----
        If the user was deleted, the object can not be used any further and
        any API call will result in an error. The key storage is also deleted.

        """
        self.__request_api("DELETE", self.user.hash)
        self.keys.delete()

    def find(self, user):
        """
        Returns the public key of an user.

        Parameters
        ----------
        param user : string
            The user name.

        Returns
        -------
        string
            PEM formatted public key.

        """
        return self.__request_api("GET", Pssst._User(user).hash + "/key")

    def pull(self):
        """
        Pulls all messages from the box.

        Returns
        -------
        list of byte strings
            The message data.

        """
        data = self.__request_api("GET", self.user.hash + "/box") or []

        return [self.keys.key.decrypt(
            _decode(message["data"]),
            _decode(message["nonce"])
        ) for message in data]

    def push(self, user, data):
        """
        Pushes a message into a box.

        Parameters
        ----------
        param user : string
            The user name.
        param data : byte string
            The message data.

        """
        user = Pssst._User(user)

        if user.name not in self.keys.list():
            self.keys.save(user.name, self.find(user.name))

        data, nonce = Pssst._Key(self.keys.load(user.name)).encrypt(data)

        self.__request_api("PUT", user.hash + "/box", {
            "nonce": _encode(nonce),
            "data": _encode(data)
        }, False)


class CLI:
    """
    Pssst CLI utility class.

    Static Methods
    --------------
    profile(username)
        Returns the profile properties.
    usage(text, *args)
        Prints the usage colored.

    """
    @staticmethod
    def profile(username="~"):
        """
        Returns the profile properties.

        Parameters
        ----------
        param username : string, optional (default is ~)
            The username or profile directory.

        Returns
        -------
        tuple
            The username, password and server.

        """
        if re.match("^[\/|\.]", username):
            path = os.path.join(os.path.abspath(username), ".pssst")
        elif username == "~":
            path = os.path.join(os.path.expanduser("~"), ".pssst")
        else:
            path = None

        if path:
            with io.open(path) as file:
                username = file.read()

        username, password, server = Pssst._User(username).profile

        if not password:
            password = getpass("Password (hidden): ")

        return (username, password, server)

    @staticmethod
    def usage(text, *args):
        """
        Prints the usage colored.

        Parameters
        ----------
        param text : string
            Usage text to print.
        param args: list of strings
            Usage parameters.

        Notes
        -----
        Color is only used on POSIX compatible systems.

        """
        for line in (text % args).split("\n")[1:-1]:
            line = line[4:]

            if os.name in ["posix"]:

                # Color description
                if re.match("^.* version \d+\.\d+\.\d+$", line):
                    line = line.replace("version", "version\x1B[34;1m")
                    line = "\x1B[39;1m%s\x1B[0m" % line

                # Color list titles
                elif re.match("^[A-Za-z ]+:$", line):
                    line = "\x1B[34m%s\x1B[0m" % line

                # Color list points
                elif re.match("^  (-.|[a-z]+)", line):
                    line = line.replace("   ", "   \x1B[37;0m")
                    line = "\x1B[34;1m%s\x1B[0m" % line

            print(line)


def main(script, command="--help", username="~", receiver=None, *message):
    """
          __________                  ___
         /  ____   /_________________/  /__
        /  /___/  / ____/ ____/ ____/  ___/
       /  _______/___  /___  /___  /  /__
      /__/      /_____/_____/_____/_____/

      CLI version %s

    Usage:
      %s [option|command] [~|username:password@server] [receiver message...]

    Options:
      -h, --help      Shows the usage
      -l, --license   Shows the license
      -v, --version   Shows the version

    Available commands:
      create   Create user
      delete   Delete user
      pull     Pull messages
      push     Push message

    Report bugs to <christian@uhsat.de>
    """
    try:
        if username:
            pssst = Pssst(*CLI.profile(username))

        if command in ("/?", "-h", "--help", "help"):
            CLI.usage(main.__doc__, __version__, os.path.basename(script))

        elif command in ("-l", "--license"):
            print(__doc__.strip())

        elif command in ("-v", "--version"):
            print("Pssst CLI " + __version__)

        elif command in ("--create", "create") and username:
            pssst.create()
            print("Created %s" % pssst.user)

        elif command in ("--delete", "delete") and username:
            pssst.delete()
            print("Deleted %s" % pssst.user)

        elif command in ("--pull", "pull") and username:
            for data in pssst.pull():
                print(data.decode("utf-8"))

        elif command in ("--push", "push") and username and receiver:
            pssst.push(receiver, " ".join(message))
            print("Message send")

        else:
            print("Unknown command or invalid username: " + command)
            print("Please use --help for help on usage.")
            return 2 # Incorrect usage

    except KeyboardInterrupt:
        return "Abort"

    except ConnectionError:
        return "Error: Connection failed"

    except Timeout:
        return "Error: Connection timeout"

    except Exception as ex:
        return "Error: %s" % ex


if __name__ == "__main__":
    sys.exit(main(*sys.argv))
